geode-core/src/main/java/org/apache/geode/cache/query/internal/index/IndexManager.java

```diff
@@ -181,11 +181,15 @@ public IndexManager(InternalCache cache, Region region) {
    * aggressively. But the large hiccup will eventually be rolled off as time is always increasing
    * This is a fix for #47475
    *
-   * @param operationTime the last modified time from version tag
+   * @param lastModifiedTime the last modified time from version tag
    */
-  public static boolean setIndexBufferTime(long operationTime, long currentCacheTime) {
-    long timeDifference = currentCacheTime - operationTime;
-    return setNewLargestValue(SAFE_QUERY_TIME, currentCacheTime + timeDifference);
+  public static void setIndexBufferTime(long lastModifiedTime, long currentCacheTime) {
+    long oldValue = SAFE_QUERY_TIME.get();
+    long newValue = currentCacheTime + currentCacheTime - lastModifiedTime;
+    if (oldValue < newValue) {
+      // use compare and set in case the value has been changed since we got the old value
+      SAFE_QUERY_TIME.compareAndSet(oldValue, newValue);
```

Comment:
What happens if this compareAndSet fails but we still have the largest safe query time?  Shouldn't there be a retry?

Reply:
I am just simplifying the old code. Based on my understanding of the old way, it didn't do any retry.

Reply:
Ah, I guess that was a mistake in the original implementation.  It definitely should retry in the original instead of returning.

Reply:
would this be another ticket though?

Reply:
Sounds good, Sorry I meant to click approve.